{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TSS mapping -1\n",
    "- Add the name of the gene to the depth file\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import the required libraries\n",
    "import pandas as pd \n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Position</th>\n",
       "      <th>Depth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Position  Depth\n",
       "0         1      4\n",
       "1         2     10\n",
       "2         3     13\n",
       "3         4     14\n",
       "4         5     16\n",
       "5         6     18\n",
       "6         7     19\n",
       "7         8     22\n",
       "8         9     22\n",
       "9        10     24"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the data\n",
    "depth_KO1 = pd.read_csv(\"depth/depth_sorted_trimmed_MBDSK1-cap_S39_R1_001.txt\", delimiter = \"\\t\", names = [\"Chr\", \"Position\", \"Depth\"], usecols = [\"Position\", \"Depth\"])\n",
    "depth_KO2 = pd.read_csv(\"depth/depth_sorted_trimmed_MBDSK2-cap_S40_R1_001.txt\", delimiter = \"\\t\", names = [\"Chr\", \"Position\", \"Depth\"], usecols = [\"Position\", \"Depth\"])\n",
    "depth_KO3 = pd.read_csv(\"depth/depth_sorted_trimmed_MBDSK3-cap_S41_R1_001.txt\", delimiter = \"\\t\", names = [\"Chr\", \"Position\", \"Depth\"], usecols = [\"Position\", \"Depth\"])\n",
    "depth_WT1 = pd.read_csv(\"depth/depth_sorted_trimmed_MBWT1-cap_S36_R1_001.txt\", delimiter = \"\\t\", names = [\"Chr\", \"Position\", \"Depth\"], usecols = [\"Position\", \"Depth\"])\n",
    "depth_WT2 = pd.read_csv(\"depth/depth_sorted_trimmed_MBWT2-cap_S37_R1_001.txt\", delimiter = \"\\t\", names = [\"Chr\", \"Position\", \"Depth\"], usecols = [\"Position\", \"Depth\"])\n",
    "depth_WT3 = pd.read_csv(\"depth/depth_sorted_trimmed_MBWT3-cap_S38_R1_001.txt\", delimiter = \"\\t\", names = [\"Chr\", \"Position\", \"Depth\"], usecols = [\"Position\", \"Depth\"])\n",
    "\n",
    "depth_KO1.head(10)  # Data is properly loaded"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Normalization step as in the shortRNAs paper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Check the total depth in all samples. \n",
    "# total_depth_KO1 = depth_KO1[\"Depth\"].sum()\n",
    "# total_depth_KO2 = depth_KO2[\"Depth\"].sum()\n",
    "# total_depth_KO3 = depth_KO3[\"Depth\"].sum()\n",
    "# total_depth_WT1 = depth_WT1[\"Depth\"].sum()\n",
    "# total_depth_WT2 = depth_WT2[\"Depth\"].sum()\n",
    "# total_depth_WT3 = depth_WT3[\"Depth\"].sum()\n",
    "\n",
    "# # Create a dictionary to store the name of the data -> depth sum\n",
    "# depth_sums = {'KO1': total_depth_KO1, 'KO2': total_depth_KO2, 'KO3': total_depth_KO3, 'WT1': total_depth_WT1, 'WT2': total_depth_WT2, 'WT3': total_depth_WT3}\n",
    "\n",
    "# # Make a bar plot\n",
    "# plt.bar(depth_sums.keys(), depth_sums.values())\n",
    "# plt.title(\"Bar plot of the total depth across samples\")\n",
    "# plt.xlabel(\"Data sets\")\n",
    "# plt.ylabel(\"Depth sum\")\n",
    "# plt.show()\n",
    "\n",
    "# depth_sums.values()\n",
    "\n",
    "# # In this plot it can be clearly seen how the total depth across samples varies, that's why we need to normalize the reads. \n",
    "# # The aim is that instead of having raw depth values, we want relative RNA abundance at each position, which is: (depth_at_position_x/total_depth) * 100\n",
    "\n",
    "# depth_KO1[\"Nor_Depth\"] = (depth_KO1[\"Depth\"] / total_depth_KO1) * 100\n",
    "# depth_KO2[\"Nor_Depth\"] = (depth_KO2[\"Depth\"] / total_depth_KO2) * 100\n",
    "# depth_KO3[\"Nor_Depth\"] = (depth_KO3[\"Depth\"] / total_depth_KO3) * 100\n",
    "# depth_WT1[\"Nor_Depth\"] = (depth_WT1[\"Depth\"] / total_depth_WT1) * 100\n",
    "# depth_WT2[\"Nor_Depth\"] = (depth_WT2[\"Depth\"] / total_depth_WT2) * 100\n",
    "# depth_WT3[\"Nor_Depth\"] = (depth_WT3[\"Depth\"] / total_depth_WT3) * 100\n",
    "\n",
    "# # Check the total depth in all samples. \n",
    "# total_nor_depth_KO1 = depth_KO1[\"Nor_Depth\"].sum()\n",
    "# total_nor_depth_KO2 = depth_KO2[\"Nor_Depth\"].sum()\n",
    "# total_nor_depth_KO3 = depth_KO3[\"Nor_Depth\"].sum()\n",
    "# total_nor_depth_WT1 = depth_WT1[\"Nor_Depth\"].sum()\n",
    "# total_nor_depth_WT2 = depth_WT2[\"Nor_Depth\"].sum()\n",
    "# total_nor_depth_WT3 = depth_WT3[\"Nor_Depth\"].sum()\n",
    "\n",
    "# # Create a dictionary to store the name of the data -> depth sum\n",
    "# depth_nor_sums = {'KO1': total_nor_depth_KO1, 'KO2': total_nor_depth_KO2, 'KO3': total_nor_depth_KO3, 'WT1': total_nor_depth_WT1, 'WT2': total_nor_depth_WT2, 'WT3': total_nor_depth_WT3}\n",
    "\n",
    "# # Make a bar plot\n",
    "# plt.bar(depth_nor_sums.keys(), depth_nor_sums.values())\n",
    "# plt.title(\"Bar plot of the total normalized depth across samples\")\n",
    "# plt.xlabel(\"Data sets\")\n",
    "# plt.ylabel(\"Depth sum\")\n",
    "# plt.show()\n",
    "\n",
    "# depth_nor_sums.values()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It can be seen how now the data has been normalized and the each group have the same total of reads"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Load the annotation file and add the genes in the depth files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total length of the anotation file is 8296. \n",
      "4106 start values are duplicated. \n",
      "4106 end values are duplicated\n",
      "gene              4097\n",
      "CDS               4000\n",
      "repeat_region      107\n",
      "tRNA                45\n",
      "mobile_element      42\n",
      "rRNA                 3\n",
      "misc_RNA             2\n",
      "Name: type, dtype: int64\n",
      "Are all values filled with gene names? False\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDS</td>\n",
       "      <td>1</td>\n",
       "      <td>1524</td>\n",
       "      <td>+</td>\n",
       "      <td>dnaA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CDS</td>\n",
       "      <td>2052</td>\n",
       "      <td>3260</td>\n",
       "      <td>+</td>\n",
       "      <td>dnaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CDS</td>\n",
       "      <td>3280</td>\n",
       "      <td>4437</td>\n",
       "      <td>+</td>\n",
       "      <td>recF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CDS</td>\n",
       "      <td>4434</td>\n",
       "      <td>4997</td>\n",
       "      <td>+</td>\n",
       "      <td>MB0004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CDS</td>\n",
       "      <td>5123</td>\n",
       "      <td>7267</td>\n",
       "      <td>+</td>\n",
       "      <td>gyrB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CDS</td>\n",
       "      <td>7302</td>\n",
       "      <td>9818</td>\n",
       "      <td>+</td>\n",
       "      <td>gyrA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CDS</td>\n",
       "      <td>9914</td>\n",
       "      <td>10828</td>\n",
       "      <td>+</td>\n",
       "      <td>MB0007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>tRNA</td>\n",
       "      <td>10887</td>\n",
       "      <td>10960</td>\n",
       "      <td>+</td>\n",
       "      <td>ILET</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>tRNA</td>\n",
       "      <td>11112</td>\n",
       "      <td>11184</td>\n",
       "      <td>+</td>\n",
       "      <td>ALAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>CDS</td>\n",
       "      <td>11874</td>\n",
       "      <td>12311</td>\n",
       "      <td>-</td>\n",
       "      <td>MB0008c</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    type  start    end strand gene_name\n",
       "1    CDS      1   1524      +      dnaA\n",
       "3    CDS   2052   3260      +      dnaN\n",
       "5    CDS   3280   4437      +      recF\n",
       "7    CDS   4434   4997      +    MB0004\n",
       "9    CDS   5123   7267      +      gyrB\n",
       "11   CDS   7302   9818      +      gyrA\n",
       "13   CDS   9914  10828      +    MB0007\n",
       "15  tRNA  10887  10960      +      ILET\n",
       "17  tRNA  11112  11184      +      ALAT\n",
       "19   CDS  11874  12311      -   MB0008c"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the gff file:\n",
    "anot_file = pd.read_csv(\"Ref_genome/sequence.gff\", sep = \"\\t\", comment=\"#\", names = [\"seqid\", \"source\", \"type\", \"start\", \"end\", \"score\", \"strand\", \"phase\", \"attributes\"], skiprows=4)\n",
    "anot_file.to_excel(\"Data_outputs/raw_anot_file.xlsx\")\n",
    "\n",
    "# Check if in the column type threre are rows that are duplicated:\n",
    "print(f\"The total length of the anotation file is {len(anot_file)}. \\n{anot_file['start'].duplicated().sum()} start values are duplicated. \\n{anot_file['end'].duplicated().sum()} end values are duplicated\")\n",
    "\n",
    "# We see that not all rows are duplicated\n",
    "print(anot_file[\"type\"].value_counts())  # see there are other factors, which can have other positions w.r.t to the gene \n",
    "\n",
    "# In this experiment we are going to keep the CDS and the locus tag and genes associated with those. \n",
    "anot_file = anot_file[anot_file[\"type\"] != \"gene\"]\n",
    "# anot_file.head(10)\n",
    "\n",
    "# Extract gene name if present, otherwise extract locus tag\n",
    "anot_file[\"gene_name\"] = anot_file[\"attributes\"].str.extract(r'gene=([^;]+);', expand=False)\n",
    "\n",
    "# If gene_name is NaN, use the locus tag instead\n",
    "anot_file[\"gene_name\"] = anot_file[\"gene_name\"].fillna(anot_file[\"attributes\"].str.extract(r'locus_tag=([^;]+)', expand=False))\n",
    "\n",
    "# Remove the part of the locus_tag starting with 'BQ2027_'\n",
    "anot_file['gene_name'] = anot_file['gene_name'].str.replace(r'BQ2027_', '', regex=True)\n",
    "\n",
    "print(f\"Are all values filled with gene names? {anot_file['gene_name'].notna().all()}\")\n",
    "\n",
    "# Drop non-important columns:\n",
    "anot_file = anot_file.drop(columns=[\"seqid\", \"source\", \"score\", \"phase\", \"attributes\"])\n",
    "\n",
    "# Get an excel file\n",
    "# anot_file = anot_file.drop(anot_file.columns[0], axis = 1)\n",
    "anot_file.to_excel(\"Data_outputs/anot_file.xlsx\")\n",
    "\n",
    "anot_file.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_gene_names(data, anot_file):\n",
    "    \"\"\"Return the depth data set with the genes\"\"\"\n",
    "    \n",
    "    # Check if all start & end positions are in the depth file \n",
    "    # print(anot_file[\"start\"].isin(data[\"Position\"]).all())\n",
    "    # print(anot_file[\"end\"].isin(data[\"Position\"]).all())\n",
    "\n",
    "    # Add the gene name for the starting position:\n",
    "    # Set the start points of the annotation file as indexes. \n",
    "    anot_file_sta_indexes = anot_file.set_index(\"start\")\n",
    "    anot_file_end_indexes = anot_file.set_index(\"end\")\n",
    "    # Extract the start positions as a list:\n",
    "    start_pos = anot_file[\"start\"].to_list()\n",
    "    end_pos = anot_file[\"end\"].to_list()\n",
    "    \n",
    "    for start_val in start_pos:\n",
    "        data.loc[start_val -1, \"Gene_name\"] = anot_file_sta_indexes.loc[start_val, \"gene_name\"] + \"_start\"\n",
    "    \n",
    "    for end_val in end_pos:\n",
    "        data.loc[end_val -1, \"Gene_name\"] = anot_file_end_indexes.loc[end_val, \"gene_name\"] + \"_end\"\n",
    "\n",
    "    for _, row in anot_file.iterrows():  \n",
    "        # print(row[\"start\"])\n",
    "        # Assign gene name for positions between start and end (inclusive)\n",
    "        data.loc[(data[\"Position\"] > row[\"start\"]) & (data[\"Position\"] < row[\"end\"]), \"Gene_name\"] = row[\"gene_name\"]\n",
    "        \n",
    "    data[\"Gene_name\"] = data[\"Gene_name\"].fillna(\"no_gene\")    \n",
    "\n",
    "    return data\n",
    "\n",
    "gene_depth_KO1 = add_gene_names(depth_KO1, anot_file)\n",
    "gene_depth_KO1.to_csv(\"Data_outputs/Depth_and_genes/genes_depth_KO1.txt\", index = False)\n",
    "gene_depth_KO2 = add_gene_names(depth_KO2, anot_file)                                              ## Run this when the TSS functions work properly\n",
    "gene_depth_KO2.to_csv(\"Data_outputs/Depth_and_genes/genes_depth_KO2.txt\", index = False)\n",
    "gene_depth_KO3 = add_gene_names(depth_KO3, anot_file)\n",
    "gene_depth_KO3.to_csv(\"Data_outputs/Depth_and_genes/genes_depth_KO3.txt\", index = False)\n",
    "gene_depth_WT1 = add_gene_names(depth_WT1, anot_file)\n",
    "gene_depth_WT1.to_csv(\"Data_outputs/Depth_and_genes/genes_depth_WT1.txt\", index = False)\n",
    "gene_depth_WT2 = add_gene_names(depth_WT2, anot_file)\n",
    "gene_depth_WT2.to_csv(\"Data_outputs/Depth_and_genes/genes_depth_WT2.txt\", index = False)\n",
    "gene_depth_WT3 = add_gene_names(depth_WT3, anot_file)\n",
    "gene_depth_WT3.to_csv(\"Data_outputs/Depth_and_genes/genes_depth_WT3.txt\", index = False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py3.9 (quant)",
   "language": "python",
   "name": "quant"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
